cpp/src/arrow/util/decimal_internal.h

```diff
@@ -442,8 +516,9 @@ struct RealTraits<float> {
   static constexpr float two_to_192(float x) { return x == 0 ? 0 : kFloatInf; }
 
   static constexpr int kMantissaBits = 24;
-  // ceil(log10(2 ^ kMantissaBits))
-  static constexpr int kMantissaDigits = 8;
+  // log10(2 ^ kMantissaBits) ~= 7.2, let's be conservative to ensure more accuracy
+  // with our conversions for Decimal values
```

Comment:
Yea, we had a persistent off-by-one rounding issue with Decimal32 and Decimal64 for the `FromReal` tests which was fixed by this, and had no detrimental effect on Decimal128/256

Reply:
As I mentioned elsewhere, float->Decimal32 probably needs to fall back on the approx algorithm.

I'm a bit surprised for Decimal64, though. Is the RoundedRightShift right?

Reply:
I reran the tests to confirm, and yea it's only the Decimal32 case which runs into this problem (and only for Float since the double case falls back to the approx algorithm). 

In the float case, Decimal32 won't hit the condition to fall back on the approx algorithm as MaxPrecision == 9 and MantissaDigits is 8 (or 7 if I keep this change). In either case, using 7 for the mantissa digits is likely more accurate given that log10(2 ^ 24) is ~7.2 as opposed to doubles which uses 16 because log(2 ^ 53) ~= 15.9 so it makes sense to ceiling that to 16.

The only alternative would be to just have Decimal32 *always* go to the approx algorithm even for the float case despite maxprecision being larger than the mantissa digits.

Reply:
I've switched this back to 8 and just unconditionally send Decimal32 to the approx algorithm. Let me know if this is sufficient or if I should put it back.